specs/ServerCertificate.md

```diff
@@ -0,0 +1,299 @@
+Server Certificate API
+===
+# Background
+
+The WebView2 team has been asked for an API to intercept when WebView2 cannot verify a server's digital certificate while loading a web page.
+This API provides an option to trust the server's TLS certificate at the application level and render the page without prompting the user about the TLS error or can cancel the request.
+
+# Description
+
+We propose adding `ReceivingServerCertificateError` API that allows you to verify TLS certificates with errors, and either continue the request
```

Comment:
I would not be averse to calling this `ServerCertificateErrorDetected`. It still makes sense to have Handled/Cancel for a "Detected" event, because you're controlling what the system does as a result of the certificate error.

Handled = true means "It's all good, move along now, nothing to see here."

Cancel = true means "Give up now, do not pass Go, do not collect $200."

Neither handled nor canceled means "Do whatever you would normally do."

Oh wait, the system caches results on a per-certificate basis. That doesn't really match the Handled/Cancel pattern, because the Handled/Cancel refers to the operation, and the expectation is that a new operation will raise a new event.

Maybe the EventArgs should have a mutable property named something like Action or CertificateAction, with possible values
* ServerCertificateAction.AlwaysAllow
* ServerCertificateAction.AlwaysReject
* ServerCertificateAction.AlwaysPrompt (default)

The "Always" makes it clearer that you are saying what to do not just for this particular access, but for all future accesses.

```cpp
               // Allow navigation to a server with a TLS certificate if
               // the error status is of type `COREWEBVIEW2_WEB_ERROR_STATUS_CERTIFICATE_IS_INVALID`
               // and trusted by the host app.
               if (errorStatus == COREWEBVIEW2_WEB_ERROR_STATUS_CERTIFICATE_IS_INVALID &&
                            ValidateServerCertificate(certificate.get()))
               {
                 CHECK_FAILURE(args->put_CertificateAction(COREWEBVIEW2_SERVER_CERTIFICATE_ACTION_ALWAYS_ALLOW));
               }
               else
               {
                 // Cancel the request for other TLS certificate error types or if untrusted by the host app.
                 CHECK_FAILURE(args->put_CertificateAction(COREWEBVIEW2_SERVER_CERTIFICATE_ACTION_ALWAYS_REJECT));
               }
```

Question: If I say "AlwaysPrompt" and then the user says "Sure, go ahead", does the system cache the "go ahead" so the next time it encounters the certificate, it just proceeds without prompting? If so, then the enum should be called "Prompt" rather than "AlwaysPrompt", because "AlwaysPrompt" implies that it prompts every time.

Reply:
In previous API review ([ClientCertificateRequested](https://github.com/MicrosoftEdge/WebView2Feedback/blob/master/specs/ClientCertificate.md)) event we decided to use Handled/Cancel properties rather than using enum property. 

Yes, so when they set handled (AlwaysAllow) to true or Handled and Cancel are false (Prompt and user responds to proceed), this decision for cached.

Don't we want to do the same? @david-risney What do you think?

Reply:
In client certificates I suppose we have similar options (handled + cancel although handled is about providing client cert rather than ignoring error). Is the caching similar as well? IIRC client cert is cached per origin per session - is that right? (we should document that) 

@monica-ch, per Raymond's question, only the Handled = true, Cancel = false case is cached correct? If you cancel or perform default action its not cached, is that correct? So for Raymond's enum it would be AlwaysAllow, Reject, Prompt?

Reply:
Regarding the name of the event ReceivingServerCertificateError -> ServerCertificateErrorDetected, I like ServerCertificateErrorDetected better, although I thought we would need an -ing name since we're in the middle of deciding what to do with the error and blocking download until then. (Although again otherwise the name ServerCertificateErrorDetected sounds much better to me)

Reply:
That's the beauty of calling it Detected. The detection has already happened. You can't stop it from happening. But what you *can* do is influence what the system does after detection. We use this pattern in many types, e.g. `ContextLinkInvoked` or `OcclusionsChanged`. The invoke or change has already occurred, but you can influence how the system responds to it. (We use this pattern a lot when the verb is `Requested`. The request has already happened. This is your chance to influence what the system does in response to the request.)

This is different from using a verb that describes the operation in progress. `Navigated` means that the navigation is already over. `Navigating` means that it's not too late.

So we can't call this `ServerCertificateEvaluated` because the evaluation is not yet complete. Switching to a verb that is a *step* in the evaluation lets us name it after the step, rather than the larger operation.

Reply:
I was going to leave a similar comment about the -ing form.  We're not receiv_ing_ a ServerCertificateError, we already have it and now we need to decide what to do.
Verb should be after "ServerCertificateError" to align with the other EventArgs at https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core?view=webview2-dotnet-1.0.1108.44

Reply:
@david-risney Thanks for the clarification. Right, client certificate is cached per origin per session. (I'll update our documentation on this).

We are only caching when handled is true and cancel is false. AlwaysAllow, Reject, Prompt sounds good to me as well.

Reply:
Raymond didn't like using Handled/Canceled for a different reason because of the caching. Good point though that we need an -ing name when using Handled/Canceled. We can do ServerCertificateErrorDetected or BlockingDueToServerCertificateError and use the enum property instead of Handled/Canceled. 

ServerCertificateErrorDetected is easier to read imho and it starts with the more important words for this event rather than ending with them.

Reply:
So what's the current proposal/plan?

Reply:
Here's the MIDL3:

```
namespace Microsoft.Web.WebView2.Core
{
    enum CoreWebView2ServerCertificateErrorAction
    {
        Reject, // Default. For page navigations shows certificate error page with 'allow anyway' opt-in button
        AlwaysAllow, // Cached for this origin+cert for this session. Cleared by calling Clear...Async
        Cancel
    }

    runtimeclass CoreWebView2ServerCertificateErrorDetectedEventArgs
    {
        CoreWebView2WebErrorStatus ErrorStatus { get; };
        String RequestUri { get; };
        CoreWebView2Certificate ServerCertificate { get; };
        CoreWebView2ServerCertificateErrorAction Action { get; set; };
        Windows.Foundation.Deferral GetDeferral();
    }

    runtimeclass CoreWebView2
    {
        // ...
        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_11")]
        {
          event Windows.Foundation.TypedEventHandler<CoreWebView2, CoreWebView2ServerCertificateErrorDetectedEventArgs> ServerCertificateErrorDetected;
          Windows.Foundation.IAsyncAction ClearServerCertificateErrorActionsAsync();
 
        }
    }
}
```

Reply:
Looks great to me. The only confusing part is the name `Reject`, because that doesn't communicate "do whatever you think is a good idea". I think the most common pattern for that is `Default`.

Reply:
Ok. I wasn't sure how to name that exactly. `Default` sounds good. I've updated the above comment